PR for Week 4 peer review

amosting opened the pull request.

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

amosting committed:
Level 0. Rename, Greet, Exit.
Deleted Duke.java, replaced with Jarvis.java and edited Jarvis.java

amosting committed:
Level 1. Echo
Added logic to echo user commands.
If user enters anything other than bye, echo. If user enters bye, reply with goodbye message.

amosting committed:
Level 2. Add, List.
Added logic to store tasks and print out a list of tasks when the request "list" is given.

amosting committed:
Level 3. Mark as Done.
Added Task.java.
Added logic for "mark" and "unmark"

amosting committed:
Level 4. ToDos, Events, Deadlines
Added Todo.java, Event.java and Dealine.java which inherit from Task.java
Added logic in Jarvis.java to handle "todo", "event" and "deadline" commands

amosting committed:
A-TextUiTesting: Automated Text UI Testing
Added command line inputs into input.txt and expected outputs into expected.txt.
Added simple logic to handle erroneous inputs, this is not completed.
Removed ASCII art because it was causing too many issues with the input output comparison.

amosting committed:
Level 5. Handle Errors.
Added JarvisException.java which is the parent of IncorrectJarvisCommandException.java, InvalidTaskNumberException.java and WrongJarvisCommandFormatException.java
Added logic to handle errors in Jarvis.java

amosting committed:
Level 6. Delete.
Added logic to handle delete command.
Added error handling for delete command.
Organised error handling to become more modular and added regex matching.

amosting committed:
Add Level 7. Save.
Added save logic to allow for task list to be saved to local drive.
Used gitignore on data.txt so that git will not track changes to data.txt as it is the file that stores the task list data.
Further modularised and cleaned up code in main.

amosting committed:
Revert "Add Level 7. Save."

This reverts commit 4aea4a8f3eb100aec0be67c8e19fd8534808b43a.

amosting committed:
Revert "Revert "Add Level 7. Save.""

This reverts commit 8cfc1a6998e54913dce554928388c4da99f6dc4a.

amosting committed:
Add Level 8. Dates and Times.
Added logic in Deadline class to handle date recognition.
Added empty function in Task class for inheritance in Deadline class.

amosting committed:
Add TaskList.java
TaskList.java added to extract out tasklist related code from main to keep code modular.
appendTask, addTask, deleteTask, getTask, countTask, isValidTaskNumber, printTask, printDelete, printTaskList all added into TaskList.java

amosting committed:
Add Storage.java
Storage.java added to extract out loading and saving file related code from main to keep code modular.
load, deleteContents, save methods added
Logic for instantiating TaskList object from a saved file also added

amosting committed:
Add Parser and Ui class and changed Jarvis.java
Parser.java added to process user commands and execute the proper actions needed, this logic was shifted from Jarvis.java for OOP
Ui.java added to handle UI interfacing with users, printing the proper information needed for a pleasant user experience, this logic was shifted form Jarvis.java for OOP
Jarvis.java edited to handle the increase in OOP

amosting committed:
A-Packages. Organize into Packages.
Organised all .java files into the Jarvis Package

amosting committed:
Revert "A-Packages. Organize into Packages."

This reverts commit 48cf1ab3807faca793ffe3ca94ce6a8ddece579f.

amosting committed:
A-Packages. Organise into packages.
Organised classes into Jarvis package for better organisation

amosting committed:
Merge commit 'ed6d4d25bd4e34461919ffe785a6f10f4c087552' into A-Gradle

amosting committed:
Add A-Gradle. Use Gradle.
Changed the file name in the Gradle build file so that gradle can run the program.

amosting committed:
A-JUnit.  Add JUnit Tests.
Added ToDoTest and DeadlineTest classes and used JUnit to test the toString() methods ToDo and Deadline classes respectively.

amosting committed:
A-Jar. Create a JAR File.
Added archiveFileName in order for .jar file to be named correctly

amosting committed:
Add Level-9. Find.
Added findTask to TaskList class to identify matching tasks and return them in a Task ArrayList.
Added printFoundTask to Ui.java to show the user the found files.

amosting committed:
Add. A-JavaDoc. JavaDoc.
Added JavaDoc descriptions to the Deadline, Event, Jarvis, Parser, Task, TaskList and ToDo classes.

amosting committed:
Merge branch 'Level-9'

ivanleekk commented on src/main/java/Jarvis/Parser.java:
```diff
@@ -0,0 +1,215 @@
+    /**
+     * Adds the list of valid commands into the validCommands ArrayList.
+     */
+    private void addValidCommands() {
```
good job on naming the method as a verb and writing it in camelCase, keep it up!

ivanleekk commented on src/main/java/Jarvis/Parser.java:
```diff
@@ -0,0 +1,215 @@
+        boolean uncheckMatch = inputCommand.matches("unmark \\d+");
+        boolean todoMatch = inputCommand.matches("todo .+");
+        boolean deadlineMatch = inputCommand.matches("deadline .+ /.+");
+        boolean eventMatch = inputCommand.matches("event .+ /.+ /.+");
```
if possible try to use a prefix like is, has or was to make it clearer that it is a boolean

awhb commented on src/main/java/Jarvis/TaskList.java:
```diff
@@ -0,0 +1,105 @@
+            try {
+                throw new InvalidTaskNumberException("Apologies Sir, the task number you provided is out of range.");
+            } catch (InvalidTaskNumberException e) {
+                System.out.println(line);
+                System.out.println(e.getMessage());
+                System.out.println("    There are currently " + this.taskList.size() + " tasks in the list.");
+                System.out.println(line);
```
Do consider letting your implemented UI class handle rendering the text to the user!

awhb commented on src/main/java/Jarvis/Task.java:
```diff
@@ -0,0 +1,52 @@
+ *     the basic information and operations related to a task.
+ * </p>
+ */
+public class Task {
```
Do consider adding the abstract keyword to abstract classes and methods so it is more obvious to others as well!

awhb commented on src/main/java/Jarvis/Ui.java:
```diff
@@ -0,0 +1,129 @@
+import java.util.ArrayList;
+
+public class Ui {
+    private static final String name = "Jarvis";
+    private static final String line = "____________________________________________________________";
+    private static final String greeting = "Good day Sir! I'm ";
+    private static final String question = "How can I help you today Sir?";
+    private static final String signOff = "Good bye Sir!";
+    private static final String listInforming = "Here are the tasks in your list Sir:";
+    private static final String markInforming = "Roger that Sir, I've marked this task as done:";
+    private static final String uncheckInforming = "Noted Sir, I've marked this task as NOT done yet:";
+    private static final String taskInforming = "As you please Sir, I've added the task:";
+    private static final String deleteInforming = "Alright Sir, I've removed this task";
+    private static final String findInforming = "Here are the matching tasks in your list Sir:";
```
Given that these are meant to be constants, do consider naming them using the constant names style stated in the Java coding standard for this course!
```suggestion
    private static final String NAME = "Jarvis";
    private static final String LINE = "____________________________________________________________";
    private static final String GREETING_MESSAGE = "Good day Sir! I'm ";
    private static final String QUESTION_MESSAGE = "How can I help you today Sir?";
    private static final String SIGN_OFF_MESSAGE = "Good bye Sir!";
    private static final String LIST_MESSAGE = "Here are the tasks in your list Sir:";
    private static final String MARK_MESSAGE = "Roger that Sir, I've marked this task as done:";
    private static final String UNCHECK_MESSAGE = "Noted Sir, I've marked this task as NOT done yet:";
    private static final String TASK_MESSAGE = "As you please Sir, I've added the task:";
    private static final String DELETE_MESSAGE = "Alright Sir, I've removed this task";
    private static final String FIND_MESSAGE = "Here are the matching tasks in your list Sir:";
```

awhb reviewed (commented):
Looks good 👍 
Some cosmetic changes requested.

amosting committed:
Add A-CheckStyle. Use CheckStyle.
Added CheckStyle to detect coding style violations, added logic in build.gradle to support checkstyle.
Edited existing code after adding CheckStyle make sure coding style is not violated.

ryanozx commented on src/main/java/Jarvis/Deadline.java:
```diff
@@ -0,0 +1,74 @@
+ * </p>
+ */
+public class Deadline extends Task {
+    protected String by;
```
These names seem to have the potential to confuse, perhaps a more descriptive name will make the differences between the two variables clearer?

ryanozx commented on src/main/java/Jarvis/Parser.java:
```diff
@@ -0,0 +1,222 @@
+     * @param validInputCommand the valid command which the user inputted
+     */
+    // identifies which command has wrong formatting and prints feedback to user
+    public static void isWrongFormat(String inputCommand, String validInputCommand) {
```
The method name seems to imply that it will return a boolean, perhaps a more appropriate method name could be "handleWrongFormat"?

ryanozx commented on src/main/java/Jarvis/TaskList.java:
```diff
@@ -0,0 +1,105 @@
+ * </p>
+ */
+public class TaskList {
+    private ArrayList<Task> taskList;
```
Perhaps a more appropriate name for this variable could be "tasks" to highlight that it stores a collection?

ryanozx commented on src/main/java/Jarvis/TaskList.java:
```diff
@@ -0,0 +1,105 @@
+        return isValid;
+    }
+
+    public ArrayList<Task> findTask(String keyword) {
```
Since multiple tasks can be returned, perhaps a more appropriate name could be "findTasks"?

amosting committed:
Add Level-10. GUI.
Added DialogBox, Launcher, Main and MainWindow classes to support javaFx.
Edited Jarvis, Parser, TaskList and Ui classes to support javaFx.

starrylight99 commented on src/main/java/Jarvis/Parser.java:
```diff
@@ -0,0 +1,220 @@
+        }
+        return null;
+    }
+}
```
code seems to be a little convoluted, maybe try using enums to replace the if/else statements. For exception handling, i think it would be better if you create separate classes for the different exceptions instead of handling it using if/else

amosting committed:
Add A-Assertions. Use Assertions.
Added assertions to Jarvis, Parser, Storage, TaskList, Ui classes.

amosting committed:
Removal of dead code and static keywords for Jarvis, Parser and Ui classes.

Dead code and unused variables are redundant and confuses other developers.
Wrong use of static keyword confuses others developers as class level methods and attributes are not actually meant to be as such.

Delete dead code and remove unnecessary static keywords.

So as to declutter code and fix wrong class level methods.

amosting committed:
A-CI. Set up CI.
Added gradle.yml file to set up CI for iP.

amosting committed:
Add priority attribute, methods and priority level enum to Task class and its children.

To support adding priority levels to tasks.

amosting committed:
Edit pattern recognition logic and existing functions to support additional priority value in Tasks.

The current pattern recognition logic does not support the additional priority string.
And the current constructors used to construct Task objects does not support Task classes.

New pattern recognition logic recognises when the user adds low/medium/high after typing the previous format of tasks.

amosting committed:
Add logic to handle new priority feature, for checking of correct user input.
Edited regex expression in TaskList to handle priority field.
Edited return statements in Ui.java to return correct task format for new priority feature.

amosting committed:
Add Ui screenshot of product

amosting committed:
Add User Guide.
Edited README.md to have a proper user guide for Jarvis chatbot.
Edited some small Ui class return strings to ensure proper and consistent formatting.

amosting committed:
Renamed to Ui.png

amosting committed:
Bug fixes

amosting committed:
Merge branch 'master' of https://github.com/amosting/ip
